File: Algoritmos de Ordenação/C++/radix_sort.cpp
#include "extra_functions.h"

int get_digito(int numero, int indice){
    for (int i=0; i<indice; i++)
        numero /= 10;

    return numero % 10;   
}

int get_qtd_digitos(int numero){
    int i = 0;
    
    while (numero > 0){
        numero /= 10;
        i++;
    }

    return i;
}

void counting_sort(int* array, int tamanho, int indice){
    int i; 

    int* saida = new int[tamanho]();
    int* count = new int[10]();

    for (i=0; i<tamanho; i++)
        count[get_digito(array[i], indice)] += 1;

    for (i=1; i<10; i++)
        count[i] += count[i-1];

    for (i=tamanho-1; i>=0; i--){
        saida[count[get_digito(array[i], indice)]-1] = array[i];
        count[get_digito(array[i], indice)] -= 1;
    }    

    for (i=0; i<tamanho; i++)
        *(array+i) = *(saida+i);

    delete[] saida;
    delete[] count;
}

void radix_sort(int* array, int tamanho){
    int qtd_digitos_max = get_qtd_digitos(getMaior(array, tamanho));
    
    for (int i=0; i<qtd_digitos_max; i++)
        counting_sort(array, tamanho, i);
}

int main(){

    // Criação do Array de tamanho Fixo
    int tamanho = 15;
    int* array = new int[tamanho];

    // Preenchimento do array com números inteiros
    preencherArray(array, tamanho);

    printArray(array, tamanho);
    radix_sort(array, tamanho);
    printArray(array, tamanho);

    delete[] array;

    return 0;
}

File: Algoritmos de Ordenação/Python/radix_sort.py
def countsort(lista, indice):
    saida = [0 for i in range(len(lista))]
    count = [0 for i in range(10)] #Precisa melhorar
    
    for num in lista:
        count[get_digito(num,indice)] += 1

    for i in range(1,len(count)):
        count[i] += count[i-1]
    
    for i in range(len(lista)-1,-1,-1):
        saida[count[get_digito(lista[i],indice)]-1] = lista[i]
        count[get_digito(lista[i],indice)] -= 1

    for i in range(len(saida)):
        lista[i] = saida[i]
    
    return

def get_digito(numero, indice):
    for i in range(indice):
        numero //= 10
    
    return numero % 10

def get_qtd_digitos(numero):
    i = 0
    
    while numero > 0:
        numero //= 10
        i += 1

    return i

def radix_sort(lista):
    qtd_digitos_max = get_qtd_digitos(max(lista))
    
    for i in range(qtd_digitos_max):
        countsort(lista,i)

    return

if __name__ == "__main__":
    from random import randint
    
    lista = [randint(0,1000) for i in range(15)]
    print(lista)
    
    radix_sort(lista)
    print(lista)